### DISCUSSION TITLE
CompleteHamiltonian2 function fails in task 2 but succeeded everywhere else

### DISCUSSION THREAD

**Discussion by Takodachi6969** [Part-II-Computational-Physics/.github#32] [General]
I have written the Hamiltonian2 like this, I am not sure it is working in task 7 but the two functions were checked to be closed. This feels like a hard code way to use H_loc and I am not sure if it is entirely correct (I don't see any point trying to write the code like this, maybe it saves some computing power but fundamentally its just the same as the pre-written function)

    def CompleteHamiltonian2(N, h=1.):

        H = np.zeros((2**N, 2**N))
        H_loc = LocalHamiltonian(h)
    
    
    
        for i in range(N-1):
            H += ft.reduce(np.kron,[np.eye(2**i),H_loc,np.eye(2**(N-2-i))])
    
        H -= ft.reduce(np.kron,[sigmaZ, np.eye(2**(N-2)),sigmaZ])
        H -= (h/2) * (np.kron(sigmaX, np.eye(2**(N-1))) + np.kron(np.eye(2**(N-1)), sigmaX))
        return H

    print(CompleteHamiltonian2(3))`

**Comment by JNHallen** [comment]
Hi! The aim of this task was not to have you improve on the pre-written function. They are indeed fundamentally the same. This task is there to get you thinking about how to construct larger matrices from local terms, with the provided function there to guide you. In the later tasks you then get to explore ways to improve on this method.

**Comment by JNHallen** [comment]
I am not sure what you mean about it not working in task 7? In task 7 you will not want to use this method, as it constructs the $2^N \times 2^N$ Hamiltonian matrix explicitly.
